nautilus_core/common/src/throttler/mod.rs

```diff
@@ -0,0 +1,277 @@
+    }
+}
+
+impl<T, F> Throttler<T, F>
+where
+    F: Fn(T) + 'static,
+    T: 'static,
+{
+    pub fn send(&self, msg: T) {
+        let throttler_clone = Throttler {
+            inner: self.inner.clone(),
+        };
+        let mut inner = self.inner.borrow_mut();
+        inner.recv_count += 1;
+
+        if inner.is_limiting || inner.delta_next() > 0 {
+            inner.limit_msg(msg, throttler_clone);
+        } else {
+            inner.send_msg(msg);
+        }
+    }
+
+    fn get_process_callback(&self) -> ThrottlerProcess<T, F> {
+        ThrottlerProcess::new(self.inner.clone())
+    }
+
+    fn get_resume_callback(&self) -> ThrottlerResume<T, F> {
+        ThrottlerResume::new(self.inner.clone())
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Tests
+////////////////////////////////////////////////////////////////////////////////
+// #[cfg(test)]
+mod tests {
+    use super::Throttler;
+
+    use rstest::*;
+
+    use crate::clock::TestClock;
+
+    // fn round_to_precision(value: f64) -> f64 {
+    //     let precision = 1;
+    //     let factor = 10f64.powi(precision as i32);
+    //     (value * factor).round() / factor
+    // }
+
+    type TestThrottler = Throttler<String, Box<dyn Fn(String)>>;
+
+    #[fixture]
+    pub fn test_throttler_buffered() -> TestThrottler {
+        let output_send: Box<dyn Fn(String)> = Box::new(|msg: String| {
+            log::debug!("Sent: {}", msg);
+        });
+        let clock = TestClock::new();
+
+        Throttler::new(
+            5,
+            1_000_000_000,
+            Box::new(clock),
+            "buffer_timer".to_string(),
+            output_send,
+            None,
+        )
+    }
+
+    #[fixture]
+    pub fn test_throttler_unbuffered() -> TestThrottler {
+        let output_drop: Box<dyn Fn(String)> = Box::new(|msg: String| {
+            log::debug!("Dropped: {}", msg);
+        });
+        let clock = TestClock::new();
+
+        Throttler::new(
+            5,
+            1_000_000_000,
+            Box::new(clock),
+            "buffer_timer".to_string(),
+            output_drop,
+            None,
+        )
+    }
+
+    #[rstest]
+    fn test_buffering_send_to_limit_becomes_throttled(
+        test_throttler_buffered: Throttler<String, Box<dyn Fn(String)>>,
+    ) {
+        let throttler = test_throttler_buffered;
+        throttler.send("MESSAGE".to_string());
+        throttler.send("MESSAGE".to_string());
+        throttler.send("MESSAGE".to_string());
+        throttler.send("MESSAGE".to_string());
+        throttler.send("MESSAGE".to_string());
+        throttler.send("MESSAGE".to_string());
+        assert_eq!(throttler.qsize(), 1);
+
+        let inner = throttler.inner.borrow();
+        assert!(inner.is_limiting);
+        assert_eq!(inner.recv_count, 6);
+        assert_eq!(inner.sent_count, 5);
+        assert_eq!(inner.clock.timer_names(), vec!["buffer_timer"]);
+    }
+
+    //     #[rstest]
```

Comment:
Next step for v1 is to implement the tests again. Some changes will have to be made to the assert logic since some of the inner fields might not be accessible. If needed making the inner fields pub is completely valid if it helps the tests going faster.

It'll also be useful to have some tests for the complex interactions between timers and callbacks. Like buffering/dropping then advancing timer and resuming and processing more messages. To advance the clock, you can create a TestClock anywhere in the code and run `advance_time` on it. Since the "time" is global in the system, it will also advance the time of the clock stored inside the throttler.

```rust
let throttler = Throtller..
// send messages
// buffering

let clock = TestClock::new()
clock.advance_time(5);

// check throttler if it has resumed or not
```